lib/OpenQA/ExpandPlaceholder.pm

```diff
@@ -0,0 +1,53 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+package OpenQA::ExpandPlaceholder;
+
+use strict;
+use warnings;
+
+# replace %NAME% with $settings{NAME}
+sub expand_placeholders {
+    my ($settings) = @_;
+
+    for my $value (values %$settings) {
+        next unless defined $value;
+
+        my %global_replaced;
+
+        eval { $value =~ s/%(\w+)%/_expand_placeholder($settings, $1, \%global_replaced)/eg; };
+        if ($@) {
+            return "Error: $@";
+        }
+    }
+    return undef;
+}
+
+sub _expand_placeholder {
+    my ($settings, $key, $global_replaced) = @_;
```

Comment:
I'd call the argument `$global_replaced` `$visited_placeholders_in_parent_scope`.